### PR TITLE
fix: asset upload cancellation

### PR DESCRIPTION
**Pull request by trOnk12**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

It looks like somehow Ktor is not doing well with cancellation of request bodies that require some writing to ByteChannel's. The problem seems to be Ktor not working well with CancelationException 

In the stacktrace, you can see, where the crash is coming from : 

image

It looks like, when building the request, Ktor lets any other exception up to thread handler expect of IOException() 

```
  @Throws(IOException::class)
  internal fun getResponseWithInterceptorChain(): Response {
    // Build a full stack of interceptors.
    val interceptors = mutableListOf<Interceptor>()
    interceptors += client.interceptors
    interceptors += RetryAndFollowUpInterceptor(client)
    interceptors += BridgeInterceptor(client.cookieJar)
    interceptors += CacheInterceptor(client.cache)
    interceptors += ConnectInterceptor
    if (!forWebSocket) {
      interceptors += client.networkInterceptors
    }
    interceptors += CallServerInterceptor(forWebSocket)

    val chain = RealInterceptorChain(
        call = this,
        interceptors = interceptors,
        index = 0,
        exchange = null,
        request = originalRequest,
        connectTimeoutMillis = client.connectTimeoutMillis,
        readTimeoutMillis = client.readTimeoutMillis,
        writeTimeoutMillis = client.writeTimeoutMillis
    )

    var calledNoMoreExchanges = false
    try {
      val response = chain.proceed(originalRequest)
      if (isCanceled()) {
        response.closeQuietly()
        throw IOException("Canceled")
      }
      return response
    } catch (e: IOException) {
      calledNoMoreExchanges = true
      throw noMoreExchanges(e) as Throwable
    } finally {
      if (!calledNoMoreExchanges) {
        noMoreExchanges(null)
      }
    }
  }
```

The issue is also described here , probably much better than above, so refer to it : 
https://github.com/awslabs/aws-sdk-kotlin/issues/733#

The fix is also taken almost straight from their PR.

Additional to overwriting OkHttpClientEngine in order to inject our own version of RequestBody()

```

class ByteChannelRequestBody(
    private val contentLength: Long?,
    private val callContext: CoroutineContext,
    private val block: () -> ByteReadChannel
) : RequestBody(), CoroutineScope {

    override val coroutineContext: CoroutineContext
        get() = callContext + producerJob + Dispatchers.IO

    private val producerJob = Job(callContext[Job])

    override fun contentType(): MediaType? = null
    override fun writeTo(sink: BufferedSink) {
        throw CancellationException()
        withJob(producerJob) {
            if (producerJob.isActive) {
                block().toInputStream().source().use {
                    sink.writeAll(it)
                }
            }
        }
    }

    override fun contentLength(): Long = contentLength ?: -1

    /**
     * Completes the given job when the block returns calling either `complete()` when the block runs
     * successfully or `completeExceptionally()` on exception.
     * @return the result of calling [block]
     */
    private inline fun <T> withJob(job: CompletableJob, block: () -> T): T {
        try {
            return block()
        } catch (ex: Exception) {
            println("Exception in withJob: $ex")
            job.completeExceptionally(ex)
            // wrap all exceptions thrown from inside `okhttp3.RequestBody#writeTo(..)` as an IOException
            // see https://github.com/awslabs/aws-sdk-kotlin/issues/733
            throw IOException(ex)
        } finally {
            job.complete()
        }
    }

}

```

We also had to do similar operation to our StreamAssetContent.kt which overrides OutgoingContent.WriteChannelContent.kt, 
which is also being used for updating the asset and also could slip the CancellationException into the thread exception handler and cause our app to crash.

```
internal class StreamAssetContent internal constructor(
    private val metadata: AssetMetadataRequest,
    private val encryptedDataSize: Long,
    private val fileContentStream: () -> Source,
    callContext: CoroutineContext,
) : OutgoingContent.WriteChannelContent(), CoroutineScope {

    private val producerJob = Job(callContext[Job])

    override val coroutineContext: CoroutineContext = callContext + producerJob

    private val openingData: String by lazy {
        val body = StringBuilder()

        // Part 1
        val strMetadata = "{\"public\": ${metadata.public}, \"retention\": \"${metadata.retentionType.name.lowercase()}\"}"

        body.append("--frontier\r\n")
        body.append("Content-Type: application/json;charset=utf-8\r\n")
        body.append("Content-Length: ")
            .append(strMetadata.length)
            .append("\r\n\r\n")
        body.append(strMetadata)
            .append("\r\n")

        // Part 2
        body.append("--frontier\r\n")
        body.append("Content-Type: application/octet-stream")
            .append("\r\n")
        body.append("Content-Length: ")
            .append(encryptedDataSize)
            .append("\r\n")
        body.append("Content-MD5: ")
            .append(metadata.md5)
            .append("\r\n\r\n")

        body.toString()
    }

    private val closingArray = "\r\n--frontier--\r\n"

    @OptIn(ExperimentalStdlibApi::class)
    override suspend fun writeTo(channel: ByteWriteChannel) {
        try {
            coroutineScope {
                if (!channel.isClosedForWrite && producerJob.isActive) {

                    channel.writeStringUtf8(openingData)
                    val contentBuffer = Buffer()
                    val fileContentStream = fileContentStream()
                    while (fileContentStream.read(contentBuffer, BUFFER_SIZE) != -1L) {
                        contentBuffer.readByteArray().let { content ->
                            channel.writePacket(ByteReadPacket(content))
                        }
                    }
                    channel.writeStringUtf8(closingArray)
                    channel.flush()
                    channel.close()
                }
            }
        } catch (e: Exception) {
            channel.flush()
            channel.close()
            producerJob.completeExceptionally(e)

            throw IOException(e.message)
        } finally {
            producerJob.complete()
        }
    }
}
```

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt [hunk 1]
```diff
@@ -125,68 +136,138 @@ internal class ScheduleNewAssetMessageUseCaseImpl(
...
 
+    // Generate the otr asymmetric key that will be used to encrypt the data
+    private fun generateOtrKey(): AES256Key {
+        return generateRandomAES256Key()
+    }
+
+    val test = CoroutineExceptionHandler() { _, throwable ->
+        kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.SEARCH)
+            .e("Error while uploading asset", throwable)
+    }
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt:177]: Not used at the moment. Leftover I assume?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt [hunk 2]
```diff
@@ -125,68 +136,138 @@ internal class ScheduleNewAssetMessageUseCaseImpl(
                 senderClientId = currentClientId,
                 status = Message.Status.PENDING,
                 editStatus = Message.EditStatus.NotEdited,
-                expectsReadConfirmation = expectsReadConfirmation
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt:128]: Asset messages don't expect read confirmation? Not so well versed about this feature but maybe it would be good if @vitorhugods could also chip in about this?

Comment (gongracr) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt:128]: If this is not needed, then we should also remove the `userPropertyRepository` from the constructor.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt [hunk 3]
```diff
@@ -125,68 +136,138 @@ internal class ScheduleNewAssetMessageUseCaseImpl(
...
+                            .collect { messageVisibility ->
+                                if (messageVisibility == MessageEntity.Visibility.DELETED) {
+                                    if (parentJob.isActive) {
+                                        parentJob.cancel()
+                                    }
+                                }
+                            }
+                    }
+
+                    kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.SEARCH)
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt:200]: Shouldn't we use `MESSAGES` as the logger scope?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt [hunk 4]
```diff
@@ -125,68 +136,138 @@ internal class ScheduleNewAssetMessageUseCaseImpl(
...
+                        .d("Uploading asset with id: ${assetMessageMetaData.assetId}")
+
+                    assetDataSource.uploadAndPersistPrivateAsset(
+                        assetMessageMetaData.mimeType,
+                        assetMessageMetaData.assetDataPath,
+                        assetMessageMetaData.otrKey,
+                        assetMessageMetaData.assetName.fileExtension()
+                    ).onFailure {
+                        kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.SEARCH)
+                            .d("Failed to upload asset with id: ${assetMessageMetaData.assetId}")
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt:210]:
```suggestion
                        kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.MESSAGES)
                            .e("Failed to upload asset with id: ${assetMessageMetaData.assetId}")
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt [hunk 5]
```diff
@@ -125,68 +136,138 @@ internal class ScheduleNewAssetMessageUseCaseImpl(
...
+                        updateAssetMessageUploadStatus(Message.UploadStatus.FAILED_UPLOAD, conversationId, message.id)
+                    }.flatMap { (assetId, sha256) ->
+                        kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.SEARCH)
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt:213]:
```suggestion
                        kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.MESSAGES)
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt [hunk 6]
```diff
@@ -125,68 +136,138 @@ internal class ScheduleNewAssetMessageUseCaseImpl(
...
+                                    assetMessageMetaData.copy(sha256Key = sha256, assetId = assetId),
+                                    Message.UploadStatus.UPLOADED
+                                )
+                            )
+                        )
+                        kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.SEARCH)
+                            .d("Updating asset message with id: ${updatedMessage.id} with assetId: $assetId")
+                        persistMessage(updatedMessage).onFailure {
+                            kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.SEARCH)
+                                .d("Failed to persist message with id: ${updatedMessage.id}")
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt:230]:
```suggestion
                            kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.MESSAGES)
                                .e("Failed to persist message with id: ${updatedMessage.id}")
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt [hunk 7]
```diff
@@ -125,68 +136,138 @@ internal class ScheduleNewAssetMessageUseCaseImpl(
...
+                            // We update the upload status to UPLOADED as the upload succeeded
+                            content = MessageContent.Asset(
+                                provideAssetMessageContent(
+                                    assetMessageMetaData.copy(sha256Key = sha256, assetId = assetId),
+                                    Message.UploadStatus.UPLOADED
+                                )
+                            )
+                        )
+                        kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.SEARCH)
+                            .d("Updating asset message with id: ${updatedMessage.id} with assetId: $assetId")
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt:227]:
```suggestion
                        kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.MESSAGES)
                            .d("Updating asset message with id: ${updatedMessage.id} with assetId: $assetId")
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt [hunk 8]
```diff
@@ -125,68 +136,138 @@ internal class ScheduleNewAssetMessageUseCaseImpl(
...
+                        persistMessage(updatedMessage).onFailure {
+                            kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.SEARCH)
+                                .d("Failed to persist message with id: ${updatedMessage.id}")
+                            // TODO: Should we fail the whole message sending if the updated message persistence fails? Check when implementing AR-2408
+                            kaliumLogger.e(
+                                "There was an error when trying to persist the updated asset message with the information returned by the backend "
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt:233]: Do we want to log messages with `withFeatureId()` as above and unify them all?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt [hunk 9]
```diff
@@ -125,68 +136,138 @@ internal class ScheduleNewAssetMessageUseCaseImpl(
...
+                            kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.SEARCH)
+                                .d("Successfully persisted message with id: ${updatedMessage.id}")
+                            // Finally we try to send the Asset Message to the recipients of the given conversation
+                            prepareAndSendAssetMessage(message, conversationId)
+                        }
+                    }
+                }
+            }
+        } catch (e: Exception) {
+            print("Exception thrown")
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt:245]: Please remove

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt [hunk 10]
```diff
@@ -125,68 +136,138 @@ internal class ScheduleNewAssetMessageUseCaseImpl(
...
+                        persistMessage(updatedMessage).onFailure {
+                            kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.SEARCH)
+                                .d("Failed to persist message with id: ${updatedMessage.id}")
+                            // TODO: Should we fail the whole message sending if the updated message persistence fails? Check when implementing AR-2408
+                            kaliumLogger.e(
+                                "There was an error when trying to persist the updated asset message with the information returned by the backend "
+                            )
+                        }.onSuccess {
+                            kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.SEARCH)
+                                .d("Successfully persisted message with id: ${updatedMessage.id}")
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt:237]:
```suggestion
                            kaliumLogger.withFeatureId(KaliumLogger.Companion.ApplicationFlow.MESSAGES)
                                .d("Successfully persisted message with id: ${updatedMessage.id}")
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt [hunk 11]
```diff
@@ -199,9 +280,16 @@ internal class ScheduleNewAssetMessageUseCaseImpl(
...
                 mimeType = mimeType,
                 metadata = when {
-                    isDisplayableImageMimeType(mimeType) && (assetHeight.isGreaterThan(0) && (assetWidth.isGreaterThan(0))) -> {
+                    isDisplayableImageMimeType(mimeType) && (
+                            assetHeight.isGreaterThan(0) && (
+                                    assetWidth.isGreaterThan(
+                                        0
+                                    )
+                                    )
+                            ) -> {
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/asset/ScheduleNewAssetMessageUseCase.kt:289]: Can we extract this to a separate function? Or format on a different way:
```suggestion
                   isDisplayableImageMimeType(mimeType) && (assetHeight.isGreaterThan(0) && (assetWidth.isGreaterThan(0))) -> {
                        AssetContent.AssetMetadata.Image(assetWidth, assetHeight)
                    }
```

This way is really hard to read.

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/v0/authenticated/AssetApiV0.kt [hunk 12]
```diff
@@ -157,19 +172,38 @@ internal class StreamAssetContent internal constructor(
...
-        channel.writeStringUtf8(closingArray)
-        channel.flush()
-        channel.close()
     }
 }
 
 private const val BUFFER_SIZE = 1024 * 8L
+
+
+
```
Comment (gongracr) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/v0/authenticated/AssetApiV0.kt:209]:
```suggestion
```

**Review by gongracr — CHANGES_REQUESTED** [review]
My understanding was that we didn't want to allow user's cancellation of any ongoing upload?
If we take this path this will introduce the need of maintenance of all the overriding classes recently added + more room for other edge cases in case Ktor behavior regarding callbacks changes in the near future for example. 

But if we still decide to allow it, I'd say I'd feel more comfortable adding some more unit tests to this PR, showing how if the call to the asset repository gets cancelled, halfway through, we will not be forwarding the error to the upper layer.

**Comment by trOnk12** [comment]
closing the PR, because of the product decision not to cancel asset upload
